File: PPUs/unit1.js
// import {TSet} from '@runtime';
console.log('Initializing unit1...');

// Uses (4):
import {TPersistent} from './classes.js'
import { symbols, set_of, property } from '@runtime';
//   - System
//   - objpas
//   - Classes
//   - sysutils

// Declared Types, excluding classes and $vmt (2):
export const TFooMode = symbols('fmSmall', 'fmMedium', 'fmLarge');
export const TFooOptions = set_of(TFooMode);

// Classes with published properties (2):
export class TFoo extends TPersistent {
      Count = property(Number, 7)
      Enabled = property(Boolean, true)
      Mode = property(TFooMode, 'fmMedium')
      Name = property(String, '')
      Options = property(TFooOptions)
}
export class TBar extends TFoo {
      Level = property(Number, 3)
      Title = property(String, '')
      Visible = property(Boolean, false)
}

File: PPUs/unit1.pas
unit Unit1;

{$mode objfpc}{$H+}{$M+}

interface

uses
  Classes, SysUtils;

type
  TFooMode = (fmSmall, fmMedium, fmLarge);
  TFooOptions = set of TFooMode;

  TFoo = class(TPersistent)
  private
    FCount: Integer;
    FEnabled: Boolean;
    FMode: TFooMode;
    FName: string;
    FOptions: TFooOptions;
  published
    property Count: Integer read FCount write FCount default 7;
    property Enabled: Boolean read FEnabled write FEnabled default True;
    property Mode: TFooMode read FMode write FMode default fmMedium;
    property Name: string read FName write FName;
    property Options: TFooOptions read FOptions write FOptions;
  end;

  TBar = class(TFoo)
  private
    FLevel: Integer;
    FTitle: string;
    FVisible: Boolean;
  published
    property Level: Integer read FLevel write FLevel default 3;
    property Title: string read FTitle write FTitle;
    property Visible: Boolean read FVisible write FVisible default False;
  end;

implementation

end.
